patchwork-registries/src/main/java/net/patchworkmc/impl/registries/ForgeModRegistryImpl.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Minecraft Forge, Patchwork Project
+ * Copyright (c) 2016-2020, 2019-2020
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.patchworkmc.impl.registries;
+
+import java.util.BitSet;
+
+import org.apache.commons.lang3.Validate;
+import com.google.common.collect.BiMap;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+import net.fabricmc.fabric.api.event.registry.RegistryEntryAddedCallback;
+
+/**
+ * This interface is for avoiding duplicated implementation as much as possible
+ * The vanilla field {@link net.minecraft.util.registry.SimpleRegistry#indexedEntries} is not used.
+ * @author Rikka0w0
+ */
+public interface ForgeModRegistryImpl<V extends IForgeRegistryEntry<V>> extends ForgeRegistryProvider, ModifiableRegistry<V> {
```

Comment:
The "entries" is a BiMap<int, Object> implemented by vanilla, it is used in the registry system and other places. It supports put(named set in the code) and clear, but it does not support remove. There is an alternative implement, use mixin to add a remove() method for that class. Or we can implement our own bimap which extends the vanilla BiHashMap, and use mixin to replace the instance of "entries". 
I do agree that the bitset can be replaced by a nextId and a list of available Ids

Reply:
In that case, why not replace it in the constructor with an instance of a class extending Int2ObjectBiMap that does support removal (As you've said)?

> Or we can implement our own bimap which extends the vanilla BiHashMap, and use mixin to replace the instance of "entries

Reply:
I use mixin to add a method to the Int2ObjectBiMap, because that seems to be the easier. We need to access some private fields, extending the class won't help. I added only 1 method, no fields. BiMap and BitSet are removed.